consensus/proto_array/src/proto_array.rs

```diff
@@ -300,6 +316,180 @@ impl ProtoArray {
+    /// The `head_block_root` is the block *root* of the latest invalid block. The
+    /// `latest_valid_hash` is a payload *hash* and should be:
+    ///
+    /// - `Some(hash)` if the block with that payload *hash* is known to be valid and is an
+    ///     ancestor of `head_block_root`.
+    /// - `None` if the latest valid ancestor of `head_block_root` is unkown.
+    ///
+    /// ## Details
+    ///
+    /// If `head_block_root` is not known to fork choice, an error is returned.
+    ///
+    /// If `latest_valid_hash` is `Some(hash)` where `hash` is either not known to fork choice
+    /// (perhaps it's junk orpre-finalization), then only the `head_block_root` block will be
+    /// invalidated (no ancestors). No error will be returned in this case.
+    ///
+    /// If `latest_valid_hash` is `Some(hash)` where `hash` is a known ancestor of
+    /// `head_block_root`, then all blocks between `head_block_root` and `latest_valid_hash` will
+    /// be invalidated. Additionally, all blocks that descend from a newly-invalidated block will
+    /// also be invalidated.
+    pub fn propagate_execution_payload_invalidation(
+        &mut self,
+        head_block_root: Hash256,
+        latest_valid_ancestor_hash: Option<ExecutionBlockHash>,
+    ) -> Result<(), Error> {
+        let mut invalidated_indices: HashSet<usize> = <_>::default();
+        let mut index = *self
+            .indices
+            .get(&head_block_root)
+            .ok_or(Error::NodeUnknown(head_block_root))?;
+
+        // Try to map the ancestor payload *hash* to an ancestor beacon block *root*.
+        let latest_valid_ancestor_root = latest_valid_ancestor_hash
+            .and_then(|hash| self.execution_block_hash_to_beacon_block_root(&hash));
+
+        // Set to `true` if both conditions are satisfied:
+        //
+        // 1. The `head_block_root` is a descendant of `latest_valid_ancestor_hash`
+        // 2. The `latest_valid_ancestor_hash` is equal to or a descendant of the finalized block.
+        let latest_valid_ancestor_is_descendant =
+            latest_valid_ancestor_root.map_or(false, |ancestor_root| {
+                self.is_descendant(ancestor_root, head_block_root)
+                    && self.is_descendant(self.finalized_checkpoint.root, ancestor_root)
+            });
+
+        // Collect all *ancestors* which were declared invalid since they reside between the
+        // `invalid_root` and the `latest_valid_ancestor_root`.
+        loop {
+            let node = self
+                .nodes
+                .get_mut(index)
+                .ok_or(Error::InvalidNodeIndex(index))?;
+
+            match node.execution_status {
+                ExecutionStatus::Valid(hash)
+                | ExecutionStatus::Invalid(hash)
+                | ExecutionStatus::Unknown(hash) => {
+                    // If we're no longer processing the `head_block_root` and the last valid
+                    // ancestor is known, exit now with an error.
```

Comment:
Should this be "unknown"?

```suggestion
                    // ancestor is unknown, exit now with an error.
```